aptos-move/block-executor/src/view.rs

```diff
@@ -627,11 +768,149 @@ impl<'a, T: Transaction, X: Executable> SequentialState<'a, T, X> {
     }
 }
 
+impl<'a, T: Transaction, X: Executable> ResourceState<T> for SequentialState<'a, T, X> {
+    fn set_base_value(&self, key: T::Key, value: ValueWithLayout<T::Value>) {
+        self.unsync_map.set_base_value(key, value);
+    }
+
+    fn read_cached_data_by_kind(
+        &self,
+        _txn_idx: TxnIndex,
+        key: &T::Key,
+        target_kind: ReadKind,
+        layout: UnknownOrLayout,
+        patch_base_value: &dyn Fn(&T::Value, Option<&MoveTypeLayout>) -> anyhow::Result<T::Value>,
+    ) -> ReadResult {
+        match self.unsync_map.fetch_data(key) {
+            Some(mut value) => {
+                // If we have a known layout, upgrade RawFromStorage value to Exchanged.
+                if let UnknownOrLayout::Known(layout) = layout {
+                    if let ValueWithLayout::RawFromStorage(v) = value {
+                        match patch_base_value(v.as_ref(), layout) {
+                            Ok(patched_value) => {
+                                let exchanged_value = ValueWithLayout::Exchanged(
+                                    Arc::new(patched_value.clone()),
+                                    layout.cloned().map(Arc::new),
+                                );
+                                self.unsync_map
+                                    .set_base_value(key.clone(), exchanged_value.clone());
+
+                                // sequential execution doesn't need to worry about concurrent change going through.
+                                value = exchanged_value;
+                            },
+                            Err(_) => {
+                                *self.incorrect_use.borrow_mut() = true;
```

Comment:
nit: this and the lower incorrect use handler ("Unsync map has RawFromStorage value type, while we are requesting value") can be re-used as a method or lambda (just needs string argument).
If used as a method could probably re-use with ParallelState implementation too.

Reply:
we need to do a pass on error handling and incorrect_use (and also being able to set speculative_error where needed , so I'm fine with doing cleanup at that point.